File: ocaml/include/graph/dijkstra.ml
type 'a edge = { s : int; t : int; cost : 'a }
type 'a graph = 'a edge list array

let add_edge g s t c =
  let e = { s = s; t = t; cost = c } in
  g.(s) <- e :: g.(s)

let dijkstra g s zero inf =
  let n = Array.length g in
  let dist = Array.make n inf in
  dist.(s) <- zero;
  let open BatHeap in
  let que = ref (of_list [(0, s)]) in
  let update v e =
    let nd = dist.(v) + e.cost in
    if dist.(e.t) > nd
    then (dist.(e.t) <- nd; que := insert !que (nd, e.t)) in
  while size !que > 0 do
    let d, v = find_min !que in
    que := del_min !que;
    if dist.(v) = d
    then List.iter (update v) g.(v)
  done; dist

File: cpp/include/graph/dijkstra.cpp
#pragma once

#include "../template/const_value.cpp"
#include "definition.cpp"

template <typename edge_t, typename cost_type = typename edge_t::cost_type>
std::vector<cost_type> dijkstra(const graph_t<edge_t> &g, int s) {
  std::vector<cost_type> d(g.size(), inf<cost_type>());
  d[s] = zero<cost_type>();
  using P = std::pair<cost_type, int>;
  std::priority_queue<P, std::vector<P>, std::greater<P>> que;
  que.push(P(zero<cost_type>(), s));
  while (!que.empty()) {
    cost_type dist = que.top().first;
    int v = que.top().second;
    que.pop();
    if (d[v] < dist) continue;
    for (const auto &e : g[v]) {
      if (d[e.to] <= d[v] + e.cost) continue;
      d[e.to] = d[v] + e.cost;
      que.push(P(d[e.to], e.to));
    }
  }
  return d;
}

template <typename edge_t, typename cost_type = typename edge_t::cost_type>
std::vector<cost_type> dijkstra(const graph_t<edge_t> &g, int s,
                                std::vector<int> &prev) {
  prev.resize(g.size());
  std::vector<cost_type> d(g.size(), inf<cost_type>());
  d[s] = zero<cost_type>();
  using P = std::pair<cost_type, int>;
  std::priority_queue<P, std::vector<P>, std::greater<P>> que;
  que.push(P(zero<cost_type>(), s));
  while (!que.empty()) {
    cost_type dist = que.top().first;
    int v = que.top().second;
    que.pop();
    if (d[v] < dist) continue;
    for (const auto &e : g[v]) {
      if (d[e.to] <= d[v] + e.cost) continue;
      d[e.to] = d[v] + e.cost;
      prev[e.to] = v;
      que.push(P(d[e.to], e.to));
    }
  }
  return d;
}
